overlord/state/state.go

```diff
@@ -482,6 +486,40 @@ func (s *State) GetMaybeTimings(timings interface{}) error {
 	return nil
 }
 
+// AddTaskStatusChangedObserver adds a callback function that will be invoked
+// whenever tasks change status.
+// NOTE: Callbacks registered this way may be invoked in the context
+// of the taskrunner, so the callbacks should be as simple as possible, and return
+// as quickly as possible, and should avoid the use of i/o code or blocking, as this
+// will stop the entire task system.
+func (s *State) AddTaskStatusChangedObserver(cb func(t *Task, old, new Status)) {
+	s.writing()
+	s.taskStatusChangedObservers = append(s.taskStatusChangedObservers, cb)
+}
+
+func (s *State) notifyTaskStatusChangedObservers(t *Task, old, new Status) {
+	for _, f := range s.taskStatusChangedObservers {
+		f(t, old, new)
```

Comment:
I cannot believe I missed this, it's so essential to have functionality for removing this! I'm not super keen on making a return value control this, especially because it's not explicit that it needs to be unregistered, I have instead added counter-parts when you register an observer, you can now unregister an observer as well.